lightning/src/ln/channel.rs

```diff
@@ -3554,73 +3567,83 @@ impl<Signer: Sign> Channel<Signer> {
-					// We got a reorg but not enough to trigger a force close, just update
-					// funding_tx_confirmed_in and return.
-					false
-				} else if self.channel_state < ChannelState::ChannelFunded as u32 {
-					panic!("Started confirming a channel in a state pre-FundingSent?: {}", self.channel_state);
-				} else {
-					// We got a reorg but not enough to trigger a force close, just update
-					// funding_tx_confirmed_in and return.
-					false
-				};
-				self.funding_tx_confirmed_in = Some(header.block_hash());
-
-				//TODO: Note that this must be a duplicate of the previous commitment point they sent us,
-				//as otherwise we will have a commitment transaction that they can't revoke (well, kinda,
-				//they can by sending two revoke_and_acks back-to-back, but not really). This appears to be
-				//a protocol oversight, but I assume I'm just missing something.
-				if need_commitment_update {
-					if self.channel_state & (ChannelState::MonitorUpdateFailed as u32) == 0 {
-						let next_per_commitment_point = self.holder_signer.get_per_commitment_point(self.cur_holder_commitment_transaction_number, &self.secp_ctx);
-						return Ok((Some(msgs::FundingLocked {
-							channel_id: self.channel_id,
-							next_per_commitment_point,
-						}), timed_out_htlcs));
-					} else {
-						self.monitor_pending_funding_locked = true;
-						return Ok((None, timed_out_htlcs));
+				for inp in tx.input.iter() {
+					if inp.previous_output == funding_txo.into_bitcoin_outpoint() {
+						log_trace!(logger, "Detected channel-closing tx {} spending {}:{}, closing channel {}", tx.txid(), inp.previous_output.txid, inp.previous_output.vout, log_bytes!(self.channel_id()));
+						return Err(msgs::ErrorMessage {
+							channel_id: self.channel_id(),
+							data: "Commitment transaction was confirmed on chain.".to_owned()
+						});
 					}
 				}
 			}
 		}
-		Ok((None, timed_out_htlcs))
+		Ok(None)
 	}
 
-	/// Called by channelmanager based on chain blocks being disconnected.
-	/// Returns true if we need to close the channel now due to funding transaction
-	/// unconfirmation/reorg.
-	pub fn block_disconnected(&mut self, header: &BlockHeader) -> bool {
-		if self.funding_tx_confirmations > 0 {
-			self.funding_tx_confirmations -= 1;
-			if self.funding_tx_confirmations == UNCONF_THRESHOLD as u64 {
-				return true;
+	/// When a new block is connected, we check the height of the block against outbound holding
+	/// cell HTLCs in case we need to give up on them prematurely and time them out. Everything
+	/// else (e.g. commitment transaction broadcasts, channel closure detection, HTLC transaction
+	/// broadcasting, etc) is handled by the ChannelMonitor.
+	///
+	/// If we return Err, the channel may have been closed, at which point the standard
+	/// requirements apply - no calls may be made except those explicitly stated to be allowed
+	/// post-shutdown.
```

Comment:
Hmm? It could be clarified, but not just `force_shutdown` is the next call, but `force_shutdown` is the *only* allowed call ever again on this `Channel` (and the stuff for `get_channel_update`). This is true in any "return channel-closing-error" case.

Reply:
I agree on the _only_ call allowed, my remark was more on the _how_ of this restriction. After this PR, if we return from `Channel::update_best_block`, we'll call `force_shutdown` in `ChannelManager::do_chain_event`.

I believe a more straightforward API would be to call `force_shutdown` directly inside `update_best_block` thus avoiding caller to forget to call `force_shutdown`. State transition to `ShutdownComplete` is only done in this latter method. 

At least comment should say, "If we return Err, `force_shutdown` must be called" ?